File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int start = 0;
        int end = 0;
        int n = s.length();
        int len = 0;
        HashSet<Character> set = new HashSet<>();

        while(end < n){
            while(set.contains(s.charAt(end))){
                set.remove(s.charAt(start));
                start++;
            }
            len = Math.max(len, (end - start)+1);
            set.add(s.charAt(end));
            end++;
        }

        return len;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        elem = set()
        start = 0
        end  = 0
        maxi = 0

        while end < len(s):
            while s[end] in elem:
                elem.remove(s[start])
                start += 1

            elem.add(s[end])
            maxi = max(maxi, end-start+1)
            end += 1

        return maxi